heed/src/reserved_space.rs

```diff
@@ -97,24 +103,33 @@ impl ReservedSpace<'_> {
 }
 
 impl io::Write for ReservedSpace<'_> {
+    #[inline]
     fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
-        if self.remaining() >= buf.len() {
-            let dest = unsafe { self.bytes.as_mut_ptr().add(self.write_head) };
-            unsafe { buf.as_ptr().copy_to_nonoverlapping(dest.cast(), buf.len()) };
-            self.write_head += buf.len();
-            self.written = usize::max(self.written, self.write_head);
-            Ok(buf.len())
-        } else {
-            Err(io::Error::from(io::ErrorKind::WriteZero))
-        }
+        self.write_all(buf)?;
+        Ok(buf.len())
     }
 
+    #[inline]
     fn write_all(&mut self, buf: &[u8]) -> io::Result<()> {
-        let count = self.write(buf)?;
-        debug_assert_eq!(count, buf.len());
+        let remaining = unsafe { self.bytes.get_unchecked_mut(self.write_head..) };
+
+        if buf.len() > remaining.len() {
+            return Err(io::Error::from(io::ErrorKind::WriteZero));
+        }
+
+        unsafe {
+            // SAFETY: we can always cast `T` -> `MaybeUninit<T>` as it's a transparent wrapper
+            let buf_uninit = std::slice::from_raw_parts(buf.as_ptr().cast(), buf.len());
+            remaining.as_mut_ptr().copy_from_nonoverlapping(buf_uninit.as_ptr(), buf.len());
+        }
+
+        self.write_head += buf.len();
+        self.written = usize::max(self.written, self.write_head);
```

Comment:
Should we ignore this possible error? Shouldn't we either assert or return an I/O error, here?

Reply:
We already check for an error further up, this is just making sure that we correctly keep track of which bytes have been initialized (`written`). 

For example:
```rust
write(7);
seek(3);
write(2);
```
Should result in `written == 7`, because we moved the `write_head` back to `3` in `seek`, and then advanced it to `5`, but the bytes in `0..7` have still been initialized, so using `max` we ensure we don’t ”unmark” bytes as initialized.

But maybe I misunderstood the question?